Include block: trigger warning on incompatible context

matej21 opened the pull request:
ref #60

dg commented:
:+1:

dg commented on tests/Latte/macros.include.escape.phpt:
```diff
@@ -0,0 +1,40 @@
+Assert::error(function () use ($latte) {
+	$latte->renderToString('<meta content="{include #foo}">{block #foo}{$value}{/block}', array('value' => 'b"ar'));
+}, E_USER_WARNING, 'Incompatible context for include, use either |escape or |noescape modifier');
+
```
I think that "use either |escape or |noescape modifier" is misleading.

matej21 replied:
any better idea?

dg replied:
`'Incompatible context for include'` seems fine.

matej21 replied:
I think it should say how to fix that.

dg replied:
How to fix it or how to bypass it?

matej21 replied:
I don't think it as a bypass. Latte just does not know how to handle it properly, so it leaves the decision to the developer.

matej21 commented on src/Latte/Macros/BlockMacros.php:
```diff
@@ -125,6 +129,13 @@ public function macroInclude(MacroNode $node, PhpWriter $writer)
 		} else {
 			$cmd = 'Latte\Macros\BlockMacrosRuntime::callBlock' . ($parent ? 'Parent' : '') . "(\$_b, $name, %node.array? + " . ($parent ? 'get_defined_vars' : '$template->getParameters') . '())';
 		}
+		$context = $this->getCompiler()->getContext() ? : array(NULL, NULL);
```
I'm not sure about this. Is there any difference between `getContext() = NULL` and `getContext() = array(NULL, NULL)`?

matej21 commented:
updated, squashed. I think it is ready to merge.

dg commented:
IMHO `{block double double}` is not very intuitive, what about `{block "double"}?

matej21 commented:
that is not very explicit. usually, you don't name a block `double` but sth like `{block foo double}`  :)

dg commented:
Of course something like `{block "foo"}`

matej21 commented:
I don't agree. I think we should use explicit syntax. Your version does not explicitly indicate that it somehow change the behaviour. It just seems like a quoted string. And some other developer unaware of this behaviour may remove it thinking that it is useless.

dg commented:
What do you don't agree with? I am not suggesting that `{block "foo"}` is good solution, it is just an idea. My point is that relation between word `double` and HTML quotes is not predictable.

matej21 commented:
This feature will not be used very often so I think we should be as explicit as possible. Maybe `{block foo doubleQuotes}`?

dg commented:
This is little bit complicated… because context is not only about escaping, it is about parsing too. The workaround is

```
<foo bar="{block foo}…{/block}">
```

dg commented on src/Latte/Macros/BlockMacros.php:
```diff
@@ -71,6 +74,7 @@ public function finalize()
 				$func = '_lb' . substr(md5($this->getCompiler()->getTemplateId() . $name), 0, 10) . '_' . preg_replace('#[^a-z0-9_]#i', '_', $name);
 				$snippet = $name[0] === '_';
 				$prolog[] = "//\n// block $name\n//\n"
+					. "\$_b->types[" . var_export($name, TRUE) . "][] = " . var_export($this->blockTypes[$name], TRUE) . ";"
```
The `[]` means, that blocks with same name can have a different contexts? Is it reasonable?

matej21 replied:
It just reflects blocks, which also use `[]`. I'm not really sure why :)

dg replied:
Blocks have more definitions (more functions), but IMHO it should have only one type.

matej21 replied:
Ok, it is probably needless. I will update it and will try do find some failing test :)

matej21 replied:
Cannot be there any issues with dynamic blocks and/or overriding block from a parent template?

dg replied:
Overriding block must have the same context as overrides, I think there is no issue…

dg commented:
Btw, I think we can always escape both quotes and remove difference between single / double / no quotes.

matej21 commented:
rebased and partially rewritten. ad block overriding - there actually was an issue, but it should be fixed now :)

I've removed that second commit. Your solution/workaround is better. And in the future we can add sth like `n:omit` which will just extract a block and omit surrounding HTML.

dg commented:
Thanks, I'll merge it today

matej21 commented:
Ok, I'll rebase once again (in an hour or so) and will update commit message and test names. Yesterday I was too focused on the functionality :)

matej21 commented:
Well.. with recent changes it has to be rewritten (almost) completely.

dg commented:
What about something like this? https://github.com/dg/latte/commits/block-type

matej21 commented:
@dg that looks great!

dg commented:
Feel free to take over and continue

matej21 commented:
rebased, renamed tests

dg commented on src/Latte/Macros/BlockMacros.php:
```diff
@@ -123,6 +127,13 @@ public function macroInclude(MacroNode $node, PhpWriter $writer)
+		$node->modifiers = preg_replace('#\|noescape\s?(?=\||\z)#i', '', $node->modifiers, -1, $found);
+		if (!$found && !preg_match('#\|?escape(?:html|htmlcomment|ical|js|url|xml)?\s?(?=\||\z)#i', $node->modifiers)) {
+			$cmd = "if (" . var_export($this->exportBlockType(), TRUE) . " !== \$_b->types[$name]) { "
+				. "trigger_error('Incompatible context for include block $destination.', E_USER_WARNING); }\n"
```
This is my fault, but maybe `for including` or `to include` is better.

dg commented:
There is the last thing I am not sure. `|noescape` is here used as `|nocontextcheck`, right? Isn't it confusing? I removed it from exception message because it is not good idea to advise people to use `noescape` at the moment they do not know what is happening.

matej21 commented:
> There is the last thing I am not sure. `|noescape` is here used as `|nocontextcheck`, right? 

yes, that is right. 

> I removed it from exception message...

maybe we can write there sth like "escape it or define it in proper context" ?

dg commented:
"define it in proper context" is good, but little bit obvious, i.e. "exception: block is in incompatible context" and "exception: block is in incompatible context, define it in proper context" give the same information.

"escape it" what do you mean?

dg commented:
what about `|nowarning`? I'll can use it in `{foreach}` too.

matej21 commented:
By "escape" I mean "use `|escape` filter"..

noescape was really misleading, nocheck is ok, but nowarning was imho better :)

dg commented:
I thinking about use nocheck as alias for (very bad name) nosafeurl, thus nocheck is better than nowarning.

matej21 commented:
ok, I will use |nocheck. What about the error message? should I include fixing possibilities (escape, nocheck) or the current message is enough?

dg commented:
How to solve context incompatibility is really complex issue, so IMHO it will be better to use simple exception message and let user to google  doc/forum…

matej21 commented:
updated and added missing https://github.com/nette/latte/pull/65/files#diff-232e670f239e959d76e63b24a5bc9cb2R54

dg commented:
Great! Can I merge it?

matej21 commented:
I think you can.

matej21 commented:
I'm an idiot. I forgot to change noescape to nocheck :)

dg commented:
I'll fix it

dg commented:
There is problem. For `<h1 n:block=title>` is context `html, tag`, because block is defined inside tag.

matej21 commented:
ok. should I send PR?

dg commented:
Done

matej21 commented:
You should also check script/style tag, which has different content type: https://github.com/nette/latte/blob/master/src%2FLatte%2FCompiler.php#L359-L363

dg commented:
Uff, it is really hard
